### PR TITLE
add child chain contracts

### PR DESCRIPTION
**Pull request by 0xAshish**

### CODE REVIEW

**Comment by atvanguard** [comment]
Should be ok to open this PR against, `dev-refactoring`

File: contracts/child/ChildToken.sol [hunk 1]
```diff
@@ -36,10 +36,10 @@ contract ChildToken is Ownable {
     uint256 output1
   );
 
-  function deposit(address user, uint256 amountOrTokenId) public onlyOwner;
+  function deposit(address user, uint256 amountOrTokenId) public;
 
   function withdraw(uint256 amountOrTokenId) public;
 
-  function setParent(address parent) public isParentOwner;
+  function setParent(address _parent) public;
```
Comment (atvanguard) [inline contracts/child/ChildToken.sol]: Since this contract `is Ownable`, the `parentOwner` variable can be removed.
What does the `parent` of the token contract mean? and why did you remove the `isParentOwner` modifier from `setParent`?

File: contracts/child/IParentToken.sol [hunk 2]
```diff
@@ -1,7 +1,7 @@
-pragma solidity ^0.4.24;
+pragma solidity ^0.5.2;
 //interface for parent contract of any child token
 
 
-interface IParentToken {
+contract IParentToken {
```
Comment (atvanguard) [inline contracts/child/IParentToken.sol]: How about we let this be an interface and make `beforeTransfer` external.

File: contracts/child/ParentTokenMock.sol [hunk 3]
```diff
@@ -1,4 +1,4 @@
-pragma solidity ^0.4.24;
+pragma solidity ^0.5.2;
```
Comment (atvanguard) [inline contracts/child/ParentTokenMock.sol]: Let's remove this mock. I trying to remove all mocks from the code. It isn't making any sense to have contract mocks.

Comment (0xAshish) [reply contracts/child/ParentTokenMock.sol]: this is an example contract for testing our pre-transfer hooks.

File: contracts/child/IParentToken.sol (hunk 2 above)
Comment (0xAshish) [reply contracts/child/IParentToken.sol]: beforeTransfer won't be called externally 😆.

File: contracts/child/ChildToken.sol (hunk 1 above)
Comment (0xAshish) [reply contracts/child/ChildToken.sol]: because abstract functions can't have modifiers.
`parent ` is for before transfer contract and `parentOwner ` is its owner.

File: contracts/child/ParentTokenMock.sol (hunk 3 above)
Comment (0xAshish) [reply contracts/child/ParentTokenMock.sol]: this is just an example contract we are using to test beforeTransfer function.

File: contracts/child/IParentToken.sol (hunk 2 above)
Comment (0xAshish) [reply contracts/child/IParentToken.sol]: can't, `beforeTransfer ` will never be called externally.

Comment (atvanguard) [reply contracts/child/IParentToken.sol]: It will be, but let's ignore it for now :)

File: contracts/child/ChildToken.sol (hunk 1 above)
Comment (atvanguard) [reply contracts/child/ChildToken.sol]: As discussed, `parentOwner` design is a little twisted, but ok for now.

File: contracts/child/IParentToken.sol (hunk 2 above)
Comment (0xAshish) [reply contracts/child/IParentToken.sol]: reverted back to interface.
